pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/uploader/kopia"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	repokeys "github.com/vmware-tanzu/velero/pkg/repository/keys"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo/service"
+)
+
+//BackupFunc mainly used to make testing more convenient
+var BackupFunc = kopia.Backup
+var RestoreFunc = kopia.Restore
+
+//kopiaProvider recorded info related with kopiaProvider
+//action which means provider handle backup or restore
+type kopiaProvider struct {
+	action         string
+	repoIdentifier string
+	bkRepo         udmrepo.BackupRepo
+	snapshotInfo   *uploader.SnapshotInfo
+	credGetter     *credentials.CredentialGetter
+	log            logrus.FieldLogger
+	uploader       *snapshotfs.Uploader
+	restoreCancel  chan struct{}
+}
+
+//NewKopiaUploaderProvider initialized with open or create a repository
+func NewKopiaUploaderProvider(
+	ctx context.Context,
+	repoIdentifier string,
+	credGetter *credentials.CredentialGetter,
+	repoKeySelector *v1.SecretKeySelector,
+	repoUID string,
+	configFile string,
+	log logrus.FieldLogger,
+	action string,
```

Comment:
This looks not comply with the principle we've discussed -- one provider instance could handle multiple Backup or Restore requests.
If we require calls to set an action, which is either `VeleroBackup` or `VeleroRestore`, the instance will bind to either backup or restore.

Reply:
`action` has been removed